src/events/tls.py

```diff
@@ -0,0 +1,172 @@
+        #         or self.charm.workload.generate_password(),
+        #     }
+        # )
+
+        csr = generate_csr(
+            private_key=self.charm.state.unit_server.private_key.encode("utf-8"),
+            subject=self.charm.state.unit_server.host,
+            sans_ip=self.charm.state.unit_server.sans.get("sans_ip", []),
+            sans_dns=self.charm.state.unit_server.sans.get("sans_dns", []),
+        )
+
+        self.charm.state.unit_server.update({"csr": csr.decode("utf-8").strip()})
+
+        self.certificates.request_certificate_creation(certificate_signing_request=csr)
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        """Handler for `certificates_available` event after provider updates signed certs."""
+        # avoid setting tls files and restarting
+        if event.certificate_signing_request != self.charm.state.unit_server.csr:
+            logger.error("Can't use certificate, found unknown CSR")
+            return
+
+        # if certificate already exists, this event must be new, flag manual restart
+        if self.charm.state.unit_server.certificate:
+            self.charm.on[f"{self.charm.restart.name}"].acquire_lock.emit()
+
+        self.charm.state.unit_server.update(
+            {"certificate": event.certificate, "ca-cert": event.ca}
+        )
+
+        self.charm.tls_manager.set_private_key()
+        self.charm.tls_manager.set_ca()
+        self.charm.tls_manager.set_certificate()
+        # self.charm.tls_manager.set_truststore()
+        # self.charm.tls_manager.set_p12_keystore()
+
+        self.charm.on[f"{self.charm.restart.name}"].acquire_lock.emit()
+
+    def _on_certificate_expiring(self, _: EventBase) -> None:
+        """Handler for `certificates_expiring` event when certs need renewing."""
+        if not (self.charm.state.unit_server.private_key or self.charm.state.unit_server.csr):
+            logger.error("Missing unit private key and/or old csr")
+            return
+
+        new_csr = generate_csr(
+            private_key=self.charm.state.unit_server.private_key.encode("utf-8"),
+            subject=self.charm.state.unit_server.host,
+            sans_ip=self.charm.state.unit_server.sans["sans_ip"],
+            sans_dns=self.charm.state.unit_server.sans["sans_dns"],
+        )
+
+        self.certificates.request_certificate_renewal(
+            old_certificate_signing_request=self.charm.state.unit_server.csr.encode("utf-8"),
+            new_certificate_signing_request=new_csr,
+        )
+
+        self.charm.state.unit_server.update({"csr": new_csr.decode("utf-8").strip()})
+
+    def _on_certificates_broken(self, _) -> None:
+        """Handler for `certificates_relation_broken` event."""
+        self.charm.state.unit_server.update({"csr": "", "certificate": "", "ca-cert": ""})
+
+        # remove all existing keystores from the unit so we don't preserve certs
+        self.charm.tls_manager.remove_stores()
```

Comment:
The comment is misleading.

If you take a look at `remove_stores`, it's `rm -fr`-ing the `.pem` and `.key` files.

(Did I understand your comment correctly?)

Reply:
Understood, how about changing the comment and renaming the method `remove_resources` or similar?

Reply:
Since we're strongly considering to use trust store / keystore, I'm actually thinking `remove_stores()` may have been still the best, "self-fullfilling prophecy" here.

Anyway, renamed to `remove_cert_files()`